rpcs3/Emu/RSX/Overlays/overlay_edit_text.cpp

```diff
@@ -28,10 +28,94 @@ namespace rsx
 				break;
 			}
 			case up:
+			{
+				if (!caret_position)
+				{
+					// Special case - caret is at the very beginning
+					break;
+				}
+				u64 current_line_start = text.rfind('\n', caret_position - 1);
+				if (current_line_start == std::string::npos)
+				{
+					// This is the first line, so we can't move upwards
+					break;
+				}
+				else
+				{
+					// current line is not first, caret should ignore '\n'
+					current_line_start += 1;
+				}
+				const u64 caret_pos_in_line = caret_position - current_line_start;
+				const u64 prev_line_end     = current_line_start - 1;
+				u64 prev_line_start         = text.rfind('\n', prev_line_end - 1); // - 1 is ignore '\n' and begin search before it
+				if (prev_line_start == std::string::npos)
+				{
+					// if previous line is the first line, there is no '\n' at the beginning, so we check 'rfind' result
+					prev_line_start = 0;
+				}
+				const u64 prev_line_length = prev_line_end - prev_line_start;
+				// TODO : Save caret position to some kind of buffer, so after switching back and forward, caret would be on initial position
+				if (prev_line_length < caret_pos_in_line)
+				{
+					// Previous line is shorter than current caret position, so caret moves to the very end of the previous line
+					caret_position = prev_line_end;
+				}
+				else
+				{
+					// Previous line is longer than current, so caret should offset from prev_line_beginning the exact same distance as in current line
+					caret_position = prev_line_start + caret_pos_in_line;
+				}
+				refresh();
+				break;
+			}
 			case down:
-				// TODO
+			{
+				if (caret_position == text.length())
+				{
+					// Special case - caret is at the very end
+					break;
+				}
+				const u64 current_line_end = text.find('\n', caret_position);
+				if (current_line_end == std::string::npos)
+				{
+					// This is the last line, so caret can't move downwards
+					break;
```

Comment:
```suggestion
					// This is the last line, so move to the end
					caret_position = text.length();
					break;
```

Reply:
so, you caret to move to the very beginning or very end, if it is on first or last line respectively?

Reply:
that's how notepad ++ does it at least.

Reply:
Yeah, that's fair. I think i've figured out a reason why index miss position. Will test now.

Reply:
https://gist.github.com/sampletext32/3575c8d7ea20a3d360a94fba9015051f
Here is your test app with my additional 'debug' feature.

Reply:
I fixed the "down" option.
Haven't looked at the debug feature yet

```
		const size_t current_line_end = text.find('\n', caret_position);
		if (current_line_end == std::string::npos)
		{
			// This is the last line, so move to the end
			caret_position = text.length();
			break;
		}
		size_t current_line_start = caret_position == 0 ? std::string::npos : text.rfind('\n', caret_position - 1);
		if (current_line_start == std::string::npos)
		{
			// If current line is the first one - there is no '\n' before it, so line starts from index 0
			current_line_start = 0;
		}
		else
		{
			// We found a newline, but we actually want the index of the next character behind the newline
			current_line_start++;
		}
		const size_t caret_pos_in_line = caret_position - current_line_start;
		const size_t next_line_start   = current_line_end + 1;
		size_t next_line_end           = text.find('\n', next_line_start);
		if (next_line_end == std::string::npos)
		{
			// If next line is the last line, there is no '\n' at the end, so we check 'find' result
			next_line_end = text.length();
		}
		// TODO : Save caret position to some kind of buffer, so after switching back and forward, caret would be on initial position
		caret_position = std::min(next_line_end, next_line_start + caret_pos_in_line);
```

Reply:
Works perfectly now with your test app.